vpn/tunnel.go

```diff
@@ -84,34 +117,60 @@ func (t *Tunnel) handleRPC(req *ManagerMessage, msgID uint64) *TunnelMessage {
 			slog.F("url", startReq.CoderUrl),
 			slog.F("tunnel_fd", startReq.TunnelFileDescriptor),
 		)
-		// TODO: actually start the tunnel
+		err := t.start(startReq)
+		var errStr string
+		if err != nil {
+			t.logger.Error(t.ctx, "failed to start tunnel", slog.Error(err))
+			errStr = err.Error()
+		}
 		resp.Msg = &TunnelMessage_Start{
 			Start: &StartResponse{
-				Success: true,
+				Success:      err == nil,
+				ErrorMessage: errStr,
 			},
 		}
 		return resp
 	case *ManagerMessage_Stop:
 		t.logger.Info(t.ctx, "stopping CoderVPN tunnel")
-		// TODO: actually stop the tunnel
-		resp.Msg = &TunnelMessage_Stop{
-			Stop: &StopResponse{
-				Success: true,
-			},
-		}
-		err := t.speaker.Close()
+		err := t.stop(msg.Stop)
+		var errStr string
 		if err != nil {
-			t.logger.Error(t.ctx, "failed to close speaker", slog.Error(err))
+			t.logger.Error(t.ctx, "failed to stop tunnel", slog.Error(err))
+			errStr = err.Error()
 		} else {
 			t.logger.Info(t.ctx, "coderVPN tunnel stopped")
 		}
+		resp.Msg = &TunnelMessage_Stop{
+			Stop: &StopResponse{
+				Success:      err == nil,
+				ErrorMessage: errStr,
+			},
+		}
 		return resp
 	default:
 		t.logger.Warn(t.ctx, "unhandled manager request", slog.F("request", msg))
 		return resp
 	}
 }
 
+func UseAsRouter() TunnelOption {
+	return func(t *Tunnel) {
+		t.router = NewRouter(t)
+	}
+}
+
+func UseAsLogger() TunnelOption {
+	return func(t *Tunnel) {
+		t.logger = t.logger.AppendSinks(t)
```

Comment:
Dangerous. I don't think we should use the tunnel itself as a sink for its own logs, since that could inadvertently lead to deadlocks if we log during a critical section of the tunnel logic.

Instead, the `Tunnel` should always log to the original logger it got on construction, and then, if configured, we should create a second logger that we pass to `NewConn`.